kafka/consumer.go

```diff
@@ -322,23 +353,97 @@ func (c *consumer) assigned(ctx context.Context, client *kgo.Client, assigned ma
 		for _, partition := range partitions {
 			tp := topicPartition{topic: topic, partition: partition}
-			pc := c.consumers[tp]
-			delete(c.consumers, tp)
-			close(pc.records)
+			if pc, ok := c.consumers[tp]; ok {
+				delete(c.consumers, tp)
+				close(pc.records)
+			}
 		}
 	}
 }
 
+// close is used on initiate clean shutdown. This call blocks until all the
+// partition consumers have processed their records and stopped.
+//
+// It holds the write lock, which cannot be acquired until the last fetch of
+// records has been sent to all the partition consumers.
+func (c *consumer) close() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for tp, pc := range c.consumers {
+		delete(c.consumers, tp)
+		close(pc.records)
+	}
+	c.wg.Wait()
+}
+
+// processFetch sends the received records for a partition to the corresponding
+// partition consumer. If topic/partition combination can't be found in the
+// consumer map, the consumer has been closed.
+//
+// It holds the consumer read lock, which cannot be acquired if the consumer is
+// closing.
+func (c *consumer) processFetch(ctx context.Context, fetches kgo.Fetches) {
+	if fetches.NumRecords() == 0 {
+		return
+	}
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	fetches.EachPartition(func(ftp kgo.FetchTopicPartition) {
+		if len(ftp.Records) == 0 {
+			return
+		}
+		tp := topicPartition{topic: ftp.Topic, partition: ftp.Partition}
+		pc, ok := c.consumers[tp]
+		if !ok {
+			// NOTE(marclop) While possible, this is unlikely to happen given the
+			// locking that's in place in the caller.
+			if c.delivery == apmqueue.AtMostOnceDeliveryType {
+				c.logWarn(errors.New(
+					"attempted to process records for revoked partition",
+				), ftp)
+			}
+			return
+		}
+		var err error
+		select {
+		// Send partition records to be processed by its dedicated goroutine.
+		case pc.records <- ftp.Records:
+			// Success.
+		case <-ctx.Done(): // Consumer.Run(ctx) context.
+			err = ctx.Err()
+		case <-c.ctx.Done(): // Forceful cancellation context.
+			err = c.ctx.Err()
+		}
+		// When using AtMostOnceDelivery, if the context is cancelled between
+		// PollRecords and this line, records will be lost.
+		if c.delivery == apmqueue.AtMostOnceDeliveryType {
+			c.logWarn(err, ftp)
+		}
+	})
+}
+
+func (c *consumer) logWarn(err error, ftp kgo.FetchTopicPartition) {
+	c.logger.Warn(
+		"data loss: failed to send records to process after commit",
+		zap.Error(err),
+		zap.String("topic", ftp.Topic),
+		zap.Int32("partition", ftp.Partition),
+		zap.Int64("offset", ftp.HighWatermark),
+		zap.Int("records", len(ftp.Records)),
+	)
+}
+
 // Implement the kgo.Hook that allows injecting the kgo.Client context that is
 // received on partition assignment.
 func (c *consumer) OnFetchRecordBuffered(r *kgo.Record) {
 	// Inject the context so records's context have cancellation.
 	if r.Context == nil {
-		c.mu.RLock()
-		defer c.mu.RUnlock()
 		// NOTE(marclop) Using the kgo.Client context may result in records
```

Comment:
We're not using it now, right? Should we update this comment?

Reply:
done